test/infrastructure/inmemory/internal/controllers/inmemorymachine_controller_test.go

```diff
@@ -0,0 +1,681 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/inmemory/api/v1alpha1"
+	cloudv1 "sigs.k8s.io/cluster-api/test/infrastructure/inmemory/internal/cloud/api/v1alpha1"
+	cmanager "sigs.k8s.io/cluster-api/test/infrastructure/inmemory/internal/cloud/runtime/manager"
+	"sigs.k8s.io/cluster-api/test/infrastructure/inmemory/internal/server"
+	"sigs.k8s.io/cluster-api/util/certs"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	secretutil "sigs.k8s.io/cluster-api/util/secret"
+)
+
+var (
+	ctx    = context.Background()
+	scheme = runtime.NewScheme()
+
+	cluster = &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+		},
+	}
+
+	cpMachine = &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "bar",
+			Labels: map[string]string{
+				clusterv1.MachineControlPlaneLabel: "",
+			},
+		},
+	}
+
+	workerMachine = &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "bar",
+		},
+	}
+)
+
+func init() {
+	_ = metav1.AddMetaToScheme(scheme)
+	_ = corev1.AddToScheme(scheme)
+	_ = cloudv1.AddToScheme(scheme)
+
+	ctrl.SetLogger(klog.Background())
+}
+
+func TestReconcileNormalCloudMachine(t *testing.T) {
+	inMemoryMachine := &infrav1.InMemoryMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "bar",
+		},
+		Spec: infrav1.InMemoryMachineSpec{
+			Behaviour: &infrav1.InMemoryMachineBehaviour{
+				VM: &infrav1.InMemoryVMBehaviour{
+					Provisioning: infrav1.CommonProvisioningSettings{
+						StartupDuration: metav1.Duration{Duration: 2 * time.Second},
+					},
+				},
+			},
+		},
+	}
+
+	t.Run("create CloudMachine", func(t *testing.T) {
+		g := NewWithT(t)
+
+		r := InMemoryMachineReconciler{
+			CloudManager: cmanager.New(scheme),
+		}
+		r.CloudManager.AddResourceGroup(klog.KObj(cluster).String())
+		c := r.CloudManager.GetResourceGroup(klog.KObj(cluster).String()).GetClient()
+
+		res, err := r.reconcileNormalCloudMachine(ctx, cluster, cpMachine, inMemoryMachine)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(res.IsZero()).To(BeFalse())
+		g.Expect(conditions.IsFalse(inMemoryMachine, infrav1.VMProvisionedCondition)).To(BeTrue())
+
+		got := &cloudv1.CloudMachine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: inMemoryMachine.Name,
+			},
+		}
+		err = c.Get(ctx, client.ObjectKeyFromObject(got), got)
+		g.Expect(err).ToNot(HaveOccurred())
+
+		t.Run("gets provisioned after the provisioning time is expired", func(t *testing.T) {
+			g := NewWithT(t)
+
+			g.Eventually(func() bool {
+				res, err := r.reconcileNormalCloudMachine(ctx, cluster, cpMachine, inMemoryMachine)
+				g.Expect(err).ToNot(HaveOccurred())
+				if !res.IsZero() {
+					time.Sleep(res.RequeueAfter / 100 * 90)
```

Comment:
We shouldn't need a sleep in an eventually? (not sure if I'm missing something)
(same below)

Reply:
it is just for avoiding unnecessary retry, but I can drop it if it is confusing

Reply:
Ah. Usually I would prefer simplicity in test code over optimization (if it's not necessary). But no strong opinions - up to you